search.py

```diff
@@ -484,6 +484,130 @@ def h(self, node):
 
         return sum(s != g for (s, g) in zip(node.state, self.goal))
 
+# ______________________________________________________________________________
+class NPuzzle(Problem):
+    """ Generalization of  the Eight Puzzle problem. The problem  consists of sliding tiles numbered from 1 to N ^2  on a NxN board,
+    where one of the squares is a blank. The state is represented as a tuple of length N^2,
+    where element at index i represents the tile number at index i , where '0' denotes empty square index
+    """
+
+
+
+    def __init__(self, initial= None, goal= None, size = 3, heuristic = 'manhattan_distance', shuffle =10):
+        """   Args:
+                            intial: intiail state of puzzle. If not passed will default to goal state
+                            size: number of rows and columns in puzzle
+                            heuristic: heuristic used in astat search
+                            shuffle: number of times to perfrom random action from the initial state.
+                    Returns:
+                        Problem instance modeling any size of sliding puzzle
+                         """
+        self.size = size
+        self.heuristic = heuristic
+        goal = goal or (tuple(range(1,size*size, 1)) + (0, ))
+        initial = initial or (tuple(range(1,size*size, 1)) + (0, ))
+        if shuffle:
+            initial = self.shuffle_state(shuffle, list(initial))
+        assert len(goal) == size*size, "length of goal tuple should be {length}".format(length=size*size)
+        assert len(initial) == size * size, "length of initial tuple should be {length}".format(length=size * size)
+        assert self.check_solvability(initial), "The puzzle is not solvable from the given initial state!"
+        Problem.__init__(self, initial, goal)
+    def shuffle_state(self, shuffle, state):
+        """ perfrom random action from the initial state"""
```

Comment:
type in the docstring here (perfrom -> perform)

Reply:
Thank you for spotting that! I'll correct that in the next commit.